src/base/settingsstorage.cpp

```diff
@@ -34,18 +36,47 @@
 
 #include "logger.h"
 #include "utils/fs.h"
-#include "settingsstorage.h"
+
+#ifdef Q_OS_MAC
+#define QSETTINGS_SYNC_IS_SAVE //whether QSettings::sync() is "atomic"
+#endif
 
 namespace
 {
-    inline QSettings *createSettings(const QString &name)
+    // Encapsulates serialization of settings in "atomic" way.
+    // write() does not leave half-written files,
+    // read() has a workaround for a case of power loss during a previous serialization
+    class TransactionalSettings
     {
+    public:
+        TransactionalSettings(const QString &name)
+            : m_name(name)
+        {
+        }
+
+        void read(QVariantHash &data);
```

Comment:
You have only one output param. Why don't you return it?
```c++
QVariantHash read();
```

It is more clearly.